pkg/auth/auth.go

```diff
@@ -0,0 +1,250 @@
+package auth
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"unicode/utf16"
+
+	_ "github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/dimchansky/utfbom"
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+)
+
+type CloudEnvironmentName string
+
+const (
+	AzurePublicCloud       CloudEnvironmentName = "AzurePublicCloud"
+	AzureUSGovernmentCloud CloudEnvironmentName = "AzureUSGovernment"
+	AzureChinaCloud        CloudEnvironmentName = "AzureChina"
+)
+
+type Config struct {
+	AKSCredential *aksCredential
+	AuthConfig    *Authentication
+	Cloud         cloud.Configuration
+	Authorizer    autorest.Authorizer
+}
+
+// getAuthorizer return autorest authorizer
+func (c *Config) getAuthorizer(resource string) (autorest.Authorizer, error) {
+	var auth autorest.Authorizer
+	var err error
+
+	var token *adal.ServicePrincipalToken
+	if c.AuthConfig.UseUserIdentity {
+		token, err = adal.NewServicePrincipalTokenFromManagedIdentity(
+			resource, &adal.ManagedIdentityOptions{ClientID: c.AuthConfig.UserIdentityClientId})
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		oauthConfig, err := adal.NewOAuthConfig(
+			c.Cloud.ActiveDirectoryAuthorityHost, c.AuthConfig.TenantID)
+		if err != nil {
+			return nil, err
+		}
+		token, err = adal.NewServicePrincipalToken(
+			*oauthConfig, c.AuthConfig.ClientID, c.AuthConfig.ClientSecret, resource)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	auth = autorest.NewBearerAuthorizer(token)
+	return auth, err
+}
+
+// SetAuthConfig sets the configuration needed for Authentication
+func (c *Config) SetAuthConfig() error {
+	var err error
+	c.Cloud = cloud.AzurePublic
+
+	if authFilepath := os.Getenv("AZURE_AUTH_LOCATION"); authFilepath != "" {
+		auth := &Authentication{}
+		err = auth.NewAuthenticationFromFile(authFilepath)
+		if err != nil {
+			return err
+		}
+		c.AuthConfig = auth
+	}
+
+	if aksCredFilepath := os.Getenv("AKS_CREDENTIAL_LOCATION"); aksCredFilepath != "" {
+		c.AKSCredential, err = NewAKSCredential(aksCredFilepath)
+		if err != nil {
+			return err
+		}
+
+		if c.AKSCredential != nil {
```

Comment:
You may not need the nil check since if err == nil, the aksCredential should not be nil.

Reply:
The old design was taken into consideration if customers want to run the provider on unmanaged AKS.

Reply:
I am checking the method "NewAKSCredential", &cred cannot be nil if err is nil.

Reply:
done.